nobrainer/validate.py

```diff
@@ -0,0 +1,199 @@
+
+
+    Returns:
+        `nibabel.spatialimages.SpatialImage`.
+    """
+    dice = np.zeros(n_classes)
+    for i in range(n_classes):
+        u = np.equal(pred, i)
+        v = np.equal(gt, i)
+        dice[i] = dice_numpy(u, v)
+
+    return dice
+
+
+def validate_from_filepaths(filepaths,
+                            predictor,
+                            block_shape,
+                            n_classes,
+                            mapping_y,
+                            return_variance=False,
+                            return_entropy=False,
+                            return_array_from_images=False,
+                            n_samples=1,
+                            normalizer=normalize_zero_one,
+                            batch_size=4,
+                            dtype=DT_X,
+                            ):
+    """Yield predictions from filepaths using a SavedModel.
+
+    Args:
+        test_csv: list, neuroimaging volume filepaths on which to predict.
+        n_classes: int, number of classifications the model is trained to output.
+        mapping_y: path-like, path to csv mapping file per command line argument.
+        block_shape: tuple of len 3, shape of blocks on which to predict.
+        predictor: TensorFlow Predictor object, predictor from previously
+            trained model.
+        block_shape: tuple of len 3, shape of blocks on which to predict.
+        normalizer: callable, function that accepts an ndarray and returns
+            an ndarray. Called before separating volume into blocks.
+        batch_size: int, number of sub-volumes per batch for prediction.
+        dtype: str or dtype object, dtype of features.
+
+    Returns:
+        None
+    """
+    for filepath in filepaths:
+
+        outputs, dice = validate_from_filepath(
+            filepath=filepath,
+            predictor=predictor,
+            n_classes=n_classes,
+            mapping_y=mapping_y,
+            block_shape=block_shape,
+            return_variance=return_variance,
+            return_entropy=return_entropy,
+            return_array_from_images=return_array_from_images,
+            n_samples=n_samples,
+            normalizer=normalizer,
+            batch_size=batch_size,
+            dtype=dtype)
+
+        outpath = Path(filepath[0])
+        suffixes = ''.join(s for s in outpath.suffixes)
+        mean_path = outpath.parent / (outpath.stem + '_mean' + suffixes)
+        variance_path = outpath.parent / \
+            (outpath.stem + '_variance' + suffixes)
+        entropy_path = outpath.parent / (outpath.stem + '_entropy' + suffixes)
+        dice_path = outpath.parent / (outpath.stem + '_dice.npy')
+        # if mean_path.is_file() or variance_path.is_file() or entropy_path.is_file():
+        #     raise Exception(str(mean_path) + " or " + str(variance_path) +
+        #                     " or " + str(entropy_path) + " already exists.")
+
+        nib.save(outputs[0], mean_path.as_posix())  # fix
```

Comment:
what should be fixed here?

Reply:
This part of the code may need to be changed, depending on how you want outputs to be saved.

Reply:
The validate command line interface now takes an "output_path" argument that lets the user set the path for the outputs of the network during validation.